xds/internal/httpfilter/rbac/rbac.go

```diff
@@ -0,0 +1,181 @@
+	"google.golang.org/protobuf/types/known/anypb"
+
+	v3rbacpb "github.com/envoyproxy/go-control-plane/envoy/config/rbac/v3"
+	rpb "github.com/envoyproxy/go-control-plane/envoy/extensions/filters/http/rbac/v3"
+)
+
+func init() {
+	httpfilter.Register(builder{})
+}
+
+type builder struct {
+}
+
+type config struct {
+	httpfilter.FilterConfig
+	config *rpb.RBAC
+}
+
+func (builder) TypeURLs() []string {
+	return []string{
+		"type.googleapis.com/envoy.extensions.filters.http.rbac.v3.RBAC",
+		"type.googleapis.com/envoy.extensions.filters.http.rbac.v3.RBACPerRoute",
+	}
+}
+
+// Parsing is the same for the base config and the override config.
+func parseConfig(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	if cfg == nil {
+		return nil, fmt.Errorf("rbac: nil configuration message provided")
+	}
+	any, ok := cfg.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("rbac: error parsing config %v: unknown type %T", cfg, cfg)
+	}
+	// All the validation logic described in A41.
+	msg := new(rpb.RBAC)
+	if err := ptypes.UnmarshalAny(any, msg); err != nil {
+		return nil, fmt.Errorf("rbac: error parsing config %v: %v", cfg, err)
+	}
+	for _, policy := range msg.GetRules().GetPolicies() {
+		// "Policy.condition and Policy.checked_condition must cause a
+		// validation failure if present." - A41
+		if policy.Condition != nil {
+			return nil, errors.New("rbac: Policy.condition is present")
+		}
+		if policy.CheckedCondition != nil {
+			return nil, errors.New("rbac: policy.CheckedCondition is present")
+		}
+
+		// "It is also a validation failure if Permission or Principal has a
+		// header matcher for a grpc- prefixed header name or :scheme." - A41
+		for _, principal := range policy.Principals {
+			if principal.GetHeader() != nil {
+				name := principal.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: principal header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+		for _, permission := range policy.Permissions {
+			if permission.GetHeader() != nil {
+				name := permission.GetHeader().GetName()
+				if name == ":scheme" || strings.HasPrefix(name, "grpc-") {
+					return nil, fmt.Errorf("rbac: permission header matcher for %v is :scheme or starts with grpc", name)
+				}
+			}
+		}
+	}
+
+	return config{config: msg}, nil
+}
+
+func (builder) ParseFilterConfig(cfg proto.Message) (httpfilter.FilterConfig, error) {
+	return parseConfig(cfg)
+}
+
+func (builder) ParseFilterConfigOverride(override proto.Message) (httpfilter.FilterConfig, error) {
+	any, ok := override.(*anypb.Any)
+	if !ok {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: unknown type %T", override, override)
+	}
+	msg := new(rpb.RBACPerRoute)
+	if err := ptypes.UnmarshalAny(any, msg); err != nil {
+		return nil, fmt.Errorf("rbac: error parsing override config %v: %v", override, err)
+	}
+	any, err := ptypes.MarshalAny(msg.Rbac)
```

Comment:
Instead of this, make `parseConfig` take `*pb.RBAC`. And in both `ParseFilterConfig` and `ParseFilterConfigOverride`, do `UnmarshalAny` then `parseConfig()`

Reply:
Done. Thanks for the suggestion.

Reply:
Done. Thanks for the suggestion.